File: plugins/geolocation/android/src/main/java/GeolocationPlugin.kt
// Copyright 2019-2023 Tauri Programme within The Commons Conservancy
// SPDX-License-Identifier: Apache-2.0
// SPDX-License-Identifier: MIT

package app.tauri.geolocation

import android.Manifest
import android.app.Activity
import android.location.Location
import android.os.Build
import android.webkit.WebView
import app.tauri.Logger
import app.tauri.PermissionState
import app.tauri.annotation.Command
import app.tauri.annotation.InvokeArg
import app.tauri.annotation.Permission
import app.tauri.annotation.PermissionCallback
import app.tauri.annotation.TauriPlugin
import app.tauri.plugin.Channel
import app.tauri.plugin.Invoke
import app.tauri.plugin.JSObject
import app.tauri.plugin.Plugin

@InvokeArg
class PositionOptions {
  var enableHighAccuracy: Boolean = false
  var maximumAge: Long = 0
  var timeout: Long = 10000
}

@InvokeArg
class WatchArgs {
    var options: PositionOptions = PositionOptions()
    lateinit var channel: Channel
}

@InvokeArg
class ClearWatchArgs {
    var channelId: Long = 0
}

// TODO: Plugin does not ask user to enable google location services (like gmaps does)

private const val ALIAS_LOCATION: String = "location"
private const val ALIAS_COARSE_LOCATION: String = "coarseLocation"

@TauriPlugin(
    permissions = [
        Permission(strings = [
            Manifest.permission.ACCESS_FINE_LOCATION,
            Manifest.permission.ACCESS_COARSE_LOCATION
        ],
            alias = ALIAS_LOCATION
        ),
        Permission(strings = [
            Manifest.permission.ACCESS_COARSE_LOCATION
        ],
            alias = ALIAS_COARSE_LOCATION
        )
    ]
)
class GeolocationPlugin(private val activity: Activity): Plugin(activity) {
    private lateinit var implementation: Geolocation
    private var watchers = hashMapOf<Long, Pair<Invoke, WatchArgs>>()

    override fun load(webView: WebView) {
        super.load(webView)
        implementation = Geolocation(activity.applicationContext)
    }

    override fun onPause() {
        super.onPause()
        // Clear all location updates on pause to avoid possible background location calls
        implementation.clearLocationUpdates()
    }

    override fun onResume() {
        super.onResume()
        // resume watchers
        for ((watcher, args) in watchers.values) {
            startWatch(watcher, args)
        }
    }

    @Command
    override fun checkPermissions(invoke: Invoke) {
        if (implementation.isLocationServicesEnabled()) {
            super.checkPermissions(invoke)
        } else {
            invoke.reject("Location services are disabled.")
        }
    }

    @Command
    override fun requestPermissions(invoke: Invoke) {
        if (implementation.isLocationServicesEnabled()) {
            super.requestPermissions(invoke)
        } else {
            invoke.reject("Location services are disabled.")
        }
    }

    @Command
    fun getCurrentPosition(invoke: Invoke) {
        val args = invoke.parseArgs(PositionOptions::class.java)

        val location = implementation.getLastLocation(args.maximumAge)
        if (location != null) {
            invoke.resolve(convertLocation(location))
        } else {
            implementation.sendLocation(args.enableHighAccuracy,
                { loc -> invoke.resolve(convertLocation(loc)) },
                { error -> invoke.reject(error) })
        }
    }

    @PermissionCallback
    private fun positionPermissionCallback(invoke: Invoke) {
        val permissionsResultJSON = JSObject()
        permissionsResultJSON.put("location", getPermissionState(ALIAS_LOCATION))
        permissionsResultJSON.put("coarseLocation", getPermissionState(ALIAS_COARSE_LOCATION))
        invoke.resolve(permissionsResultJSON)
    }

    @Command
    fun watchPosition(invoke: Invoke) {
        val args = invoke.parseArgs(WatchArgs::class.java)
        startWatch(invoke, args)
    }

    private fun startWatch(invoke: Invoke, args: WatchArgs) {
        implementation.requestLocationUpdates(
            args.options.enableHighAccuracy,
            args.options.timeout,
            { location -> args.channel.send(convertLocation(location)) },
            { error -> args.channel.sendObject(error) })

        watchers[args.channel.id] = Pair(invoke, args)
    }

    @Command
    fun clearWatch(invoke: Invoke) {
        val args = invoke.parseArgs(ClearWatchArgs::class.java)

        watchers.remove(args.channelId)

        if (watchers.isEmpty()) {
            implementation.clearLocationUpdates()
        }

        invoke.resolve()
    }

    private fun convertLocation(location: Location): JSObject {
        val ret = JSObject()
        val coords = JSObject()

        coords.put("latitude", location.latitude)
        coords.put("longitude", location.longitude)
        coords.put("accuracy", location.accuracy)
        coords.put("altitude", location.altitude)
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            coords.put("altitudeAccuracy", location.verticalAccuracyMeters)
        }
        coords.put("speed", location.speed)
        coords.put("heading", location.bearing)
        ret.put("timestamp", location.time)
        ret.put("coords", coords)

        return ret
    }
}

File: plugins/geolocation/ios/Sources/GeolocationPlugin.swift
// Copyright 2019-2023 Tauri Programme within The Commons Conservancy
// SPDX-License-Identifier: Apache-2.0
// SPDX-License-Identifier: MIT

import SwiftRs
import Tauri
import UIKit
import WebKit
import CoreLocation

class GetPositionArgs: Decodable {
  let enableHighAccuracy: Bool?
}

class WatchPositionArgs: Decodable {
  let options: GetPositionArgs
  let channel: Channel
}

class ClearWatchArgs: Decodable {
  let channelId: UInt32
}

class GeolocationPlugin: Plugin, CLLocationManagerDelegate {
  private let locationManager = CLLocationManager()
  private var isUpdatingLocation: Bool = false
  private var permissionRequests: [Invoke] = []
  private var positionRequests: [Invoke] = []
  private var watcherChannels: [Channel] = []

  override init() {
    super.init()
    locationManager.delegate = self
  }

  //
  // Tauri commands
  //

  @objc public func getCurrentPosition(_ invoke: Invoke) throws {
    let args = try invoke.parseArgs(GetPositionArgs.self)

    self.positionRequests.append(invoke)

    DispatchQueue.main.async {
      if args.enableHighAccuracy == true {
        self.locationManager.desiredAccuracy = kCLLocationAccuracyBest
      } else {
        self.locationManager.desiredAccuracy = kCLLocationAccuracyKilometer
      }

      // TODO: Use the authorizationStatus instance property with locationManagerDidChangeAuthorization(_:) instead.
      if CLLocationManager.authorizationStatus() == .notDetermined {
        self.locationManager.requestWhenInUseAuthorization()
      } else {
        self.locationManager.requestLocation()
      }
    }
  }

  @objc public func watchPosition(_ invoke: Invoke) throws {
    let args = try invoke.parseArgs(WatchPositionArgs.self)

    self.watcherChannels.append(args.channel)

    DispatchQueue.main.async {
      if args.options.enableHighAccuracy == true {
        self.locationManager.desiredAccuracy = kCLLocationAccuracyBest
      } else {
        self.locationManager.desiredAccuracy = kCLLocationAccuracyKilometer
      }

      // TODO: Use the authorizationStatus instance property with locationManagerDidChangeAuthorization(_:) instead.
      if CLLocationManager.authorizationStatus() == .notDetermined {
        self.locationManager.requestWhenInUseAuthorization()
      } else {
        self.locationManager.startUpdatingLocation()
        self.isUpdatingLocation = true
      }
    }

    invoke.resolve()
  }

  @objc public func clearWatch(_ invoke: Invoke) throws {
    let args = try invoke.parseArgs(ClearWatchArgs.self)

    self.watcherChannels = self.watcherChannels.filter { $0.id != args.channelId }

    // TODO: capacitor plugin calls stopUpdating unconditionally
    if self.watcherChannels.isEmpty {
      self.stopUpdating()
    }

    invoke.resolve()
  }

  @objc override public func checkPermissions(_ invoke: Invoke) {
    var status: String = ""

    if CLLocationManager.locationServicesEnabled() {
      // TODO: Use the authorizationStatus instance property with locationManagerDidChangeAuthorization(_:) instead.
      switch CLLocationManager.authorizationStatus() {
        case .notDetermined:
          status = "prompt"
        case .restricted, .denied:
          status = "denied"
        case .authorizedAlways, .authorizedWhenInUse:
          status = "granted"
        @unknown default:
          status = "prompt"
      }
    } else {
      invoke.reject("Location services are not enabled.")
      return
    }

    let result = ["location": status, "coarseLocation": status]

    invoke.resolve(result)
  }

  @objc override public func requestPermissions(_ invoke: Invoke) {
    if CLLocationManager.locationServicesEnabled() {
      // TODO: Use the authorizationStatus instance property with locationManagerDidChangeAuthorization(_:) instead.
      if CLLocationManager.authorizationStatus() == .notDetermined {
        self.permissionRequests.append(invoke)

        DispatchQueue.main.async {
          self.locationManager.requestWhenInUseAuthorization()
        }
      } else {
        checkPermissions(invoke)
      }
    } else {
      invoke.reject("Location services are not enabled.")
    }
  }

  //
  // Delegate methods
  //

  public func locationManager(_ manager: CLLocationManager, didFailWithError error: Error) {
    Logger.error(error)

    let requests = self.positionRequests + self.permissionRequests
    self.positionRequests.removeAll()
    self.permissionRequests.removeAll()

    for request in requests {
      request.reject(error.localizedDescription)
    }

    for channel in self.watcherChannels {
      do {
        try channel.send(error.localizedDescription)
      } catch {
        Logger.error(error)
      }
    }
  }

  public func locationManager(_ manager: CLLocationManager, didUpdateLocations locations: [CLLocation]) {
    // Respond to all getCurrentPosition() calls.
    for request in self.positionRequests {
       // The capacitor plugin uses locations.first but .last should be the most current one
       // and i don't see a reason to use old locations
       if let location = locations.last {
         let result = convertLocation(location)
         request.resolve(result)
       } else {
         request.reject("Location service returned an empty Location array.")
      }
    }

    for channel in self.watcherChannels {
      // The capacitor plugin uses locations.first but .last should be the most recent one
      // and i don't see a reason to use old locations
      if let location = locations.last {
        let result = convertLocation(location)
        do {
          try channel.send(result)
        } catch {
          Logger.error(error)
        }
      } else {
        do {
          try channel.send("Location service returned an empty Location array.")
        } catch {
          Logger.error(error)
        }
      }
    }
  }

  public func locationManager(_ manager: CLLocationManager, didChangeAuthorization status: CLAuthorizationStatus) {
    let requests = self.permissionRequests
    self.permissionRequests.removeAll()

    for request in requests {
      checkPermissions(request)
    }

    if !self.positionRequests.isEmpty {
      self.locationManager.requestLocation()
    }

    if !self.watcherChannels.isEmpty && !self.isUpdatingLocation {
      self.locationManager.startUpdatingLocation()
      self.isUpdatingLocation = true
    }
  }

  //
  // Internal/Helper methods
  //

  // TODO: Why is this pub in capacitor
  private func stopUpdating() {
    self.locationManager.stopUpdatingLocation()
    self.isUpdatingLocation = false
  }

  private func convertLocation(_ location: CLLocation) -> JsonObject {
    var ret: JsonObject = [:]
    var coords: JsonObject = [:]

    coords["latitude"] = location.coordinate.latitude
    coords["longitude"] = location.coordinate.longitude
    coords["accuracy"] = location.horizontalAccuracy
    coords["altitude"] = location.altitude
    coords["altitudeAccuracy"] = location.verticalAccuracy
    coords["speed"] = location.speed
    coords["heading"] = location.course
    ret["timestamp"] = Int((location.timestamp.timeIntervalSince1970 * 1000))
    ret["coords"] = coords

    return ret
  }
}

@_cdecl("init_plugin_geolocation")
func initPlugin() -> Plugin {
  return GeolocationPlugin()
}
